### PR TITLE
OTA-559: pkg/manifest: Add 'overrides' handling

### PR DESCRIPTION
**Pull request by wking**
Centralize this aspect of manifest filtering, so it can be shared between the cluster-version operator ([where it currently lives][1]) and `oc` (where we want [manifest extraction to understand this sort of condition][2]).

[1]: https://github.com/openshift/cluster-version-operator/blob/cce97c25292b263a5a549b2bb8ce82ed7f2d4923/pkg/cvo/sync_worker.go#L959-L987
[2]: https://issues.redhat.com/browse/OTA-559

### CODE REVIEW

File: pkg/manifest/manifest.go
Comment (petr-muller) [inline pkg/manifest/manifest.go:227]: nit: we could surface what is being overriden?

**Review by petr-muller — APPROVED** [review]
LGTM with a nit, address if needed
/hold

**Comment by petr-muller** [comment]
_grumbles about repos that do not respect gh reviews_
/lgtm

File: pkg/manifest/manifest.go
Comment (wking) [reply pkg/manifest/manifest.go:227]: [4.13.0-ec.4 CI][1]:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-sdn-serial/1631695148700143616/artifacts/e2e-aws-sdn-serial/gather-extra/artifacts/pods/openshift-cluster-version_cluster-version-operator-6f5d958c6b-k8t5l_cluster-version-operator.log | grep excluding | head -n2
I0303 16:57:25.646863       1 payload.go:210] excluding 0000_10_config-operator_01_infrastructure-TechPreviewNoUpgrade.crd.yaml group=apiextensions.k8s.io kind=CustomResourceDefinition namespace= name=infrastructures.config.openshift.io: "Default" is required, and release.openshift.io/feature-set=TechPreviewNoUpgrade
I0303 16:57:25.676524       1 payload.go:210] excluding 0000_30_cluster-api_00_credentials-request.yaml group=cloudcredential.openshift.io kind=CredentialsRequest namespace=openshift-cloud-credential-operator name=openshift-cluster-api-aws: "Default" is required, and release.openshift.io/feature-set=TechPreviewNoUpgrade
```

And that `excluding ...` context gets wrapped around in:

```console
cluster-verison-operator $ git --no-pager grep 'excluding ' pkg
pkg/payload/payload.go:                                 klog.V(2).Infof("excluding %s group=%s kind=%s namespace=%s name=%s: %v\n", manifest.OriginalFilename, manifest.GVK.Group, manifest.GVK.Kind, manifest.Obj.GetNamespace(), manifest.Obj.GetName(), err)
```

That could probably be cleaned up by leaning on `(*Manifest).String()`, but it's there.  Do we have similar wrapping around each call site?

```console
cluster-version-operator $ git --no-pager grep -A1 '[.]Include('
pkg/cvo/sync_worker.go:                 if err := task.Manifest.Include(nil, nil, nil, &capabilities, work.Overrides); err != nil {
pkg/cvo/sync_worker.go-                         klog.V(2).Infof("Skipping precreation of %s: %s", task, err)
--
pkg/cvo/sync_worker.go:                 if err := task.Manifest.Include(nil, nil, nil, &capabilities, work.Overrides); err != nil {
pkg/cvo/sync_worker.go-                         klog.V(2).Infof("Skipping %s: %s", task, err)
--
pkg/payload/payload.go:                         if err := manifest.Include(&excludeIdentifier, &requiredFeatureSet, &profile, onlyKnownCaps, nil); err != nil {
pkg/payload/payload.go-                                 klog.V(2).Infof("excluding %s group=%s kind=%s namespace=%s name=%s: %v\n", manifest.OriginalFilename, manifest.GVK.Group, manifest.GVK.Kind, manifest.Obj.GetNamespace(), manifest.Obj.GetName(), err)
```

So `Include` is covered.  What about `IncludeAllowUnknownCapabilities`?

```console
$ git --no-pager grep -A4 '[.]IncludeAllowUnknownCapabilities(' | grep -v vendor/
pkg/payload/payload.go:         updateManErr := updateManifest.IncludeAllowUnknownCapabilities(nil, nil, nil, &clusterCaps, nil, true)
pkg/payload/payload.go-
pkg/payload/payload.go-         // update manifest is enabled, no need to check
pkg/payload/payload.go-         if updateManErr == nil {
pkg/payload/payload.go-                 continue
--
pkg/payload/payload.go:                 if err := currentManifest.IncludeAllowUnknownCapabilities(nil, nil, nil, &clusterCaps, nil, true); err != nil {
pkg/payload/payload.go-                         continue
pkg/payload/payload.go-                 }
pkg/payload/payload.go-                 caps := capability.GetImplicitlyEnabledCapabilities(currentManifest.GetManifestCapabilities(),
pkg/payload/payload.go-                         updateManifest.GetManifestCapabilities(), capabilities)
--
$ git --no-pager grep '^func .*\|[.]IncludeAllowUnknownCapabilities(' pkg/payload/payload.go | grep -B1 IncludeAllowUnknownCapabilities
pkg/payload/payload.go:func GetImplicitlyEnabledCapabilities(updatePayloadManifests []manifest.Manifest, currentPayloadManifests []manifest.Manifest,
pkg/payload/payload.go:         updateManErr := updateManifest.IncludeAllowUnknownCapabilities(nil, nil, nil, &clusterCaps, nil, true)
pkg/payload/payload.go:                 if err := currentManifest.IncludeAllowUnknownCapabilities(nil, nil, nil, &clusterCaps, nil, true); err != nil {
```

So unless we wanted context logging in `GetImplicitlyEnabledCapabilities`, I think we're good.  And if we do want that context logging in `GetImplicitlyEnabledCapabilities`, we can handle that on the CVO  side without adding "here's the manifest I am" context on this side.

/hold cancel

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-sdn-serial/1631695148700143616

**Pull request merged**
